From 124ef48d20ca28f5aba91e371578a631ca6dcf8e Mon Sep 17 00:00:00 2001
From: Danesh M <daneshm90@gmail.com>
Date: Thu, 1 Mar 2018 08:54:15 -0500
Subject: [PATCH] NetD : Allow passing in interface names for wifi/data app
 restriction

CYAN-3976
CRACKLING-834

Changes from original cm-13.0 patch:

*) Introduce helper function manipulateRestrictAppsInOut() to
   reduce code duplication and cleanup the asprintf() /
   manipulateRestrictApps() call sequence.

Change-Id: I085434d70dfe00e9c27b821661fff5076d57e930
---
 server/BandwidthController.cpp | 74 ++++++++++++++++++++++++++++++++++
 server/BandwidthController.h   | 14 +++++++
 server/CommandListener.cpp     | 56 +++++++++++++++++++++++++
 3 files changed, 144 insertions(+)

diff --git a/server/BandwidthController.cpp b/server/BandwidthController.cpp
index 4962b7c..e50aa6d 100644
--- a/server/BandwidthController.cpp
+++ b/server/BandwidthController.cpp
@@ -238,6 +238,9 @@ int BandwidthController::enableBandwidthControl(bool force) {
     mGlobalAlertTetherCount = 0;
     mSharedQuotaBytes = mSharedAlertBytes = 0;
 
+    restrictAppUidsOnData.clear();
+    restrictAppUidsOnWlan.clear();
+
     flushCleanTables(false);
     std::string commands = Join(IPT_BASIC_ACCOUNTING_COMMANDS, '\n');
     return iptablesRestoreFunction(V4V6, commands, nullptr);
@@ -290,6 +293,77 @@ int BandwidthController::removeNiceApps(int numUids, char *appUids[]) {
                                  IptJumpReturn, IptOpDelete);
 }
 
+int BandwidthController::addRestrictAppsOnData(const char *iface, int numUids, char *appUids[]) {
+    return manipulateRestrictAppsInOut(iface, toStrVec(numUids, appUids), restrictAppUidsOnData,
+                                       IptOpInsert);
+}
+
+int BandwidthController::removeRestrictAppsOnData(const char *iface, int numUids, char *appUids[]) {
+    return manipulateRestrictAppsInOut(iface, toStrVec(numUids, appUids), restrictAppUidsOnData,
+                                       IptOpDelete);
+}
+
+int BandwidthController::addRestrictAppsOnWlan(const char *iface, int numUids, char *appUids[]) {
+    return manipulateRestrictAppsInOut(iface, toStrVec(numUids, appUids), restrictAppUidsOnWlan,
+                                       IptOpInsert);
+}
+
+int BandwidthController::removeRestrictAppsOnWlan(const char *iface, int numUids, char *appUids[]) {
+    return manipulateRestrictAppsInOut(iface, toStrVec(numUids, appUids), restrictAppUidsOnWlan,
+                                       IptOpDelete);
+}
+
+int BandwidthController::manipulateRestrictAppsInOut(const char *iface,
+                                                     const std::vector<std::string>& appStrUids,
+                                                     std::list<int /*appUid*/> &restrictAppUids,
+                                                     IptOp op) {
+    char *chain;
+    if (asprintf(&chain, "INPUT -i %s", iface) < 0) {
+        return -1;
+    }
+    int ret = manipulateRestrictApps(appStrUids, chain, restrictAppUids, op);
+    free(chain);
+    if (ret != 0) {
+        return ret;
+    }
+    if (asprintf(&chain, "OUTPUT -o %s", iface) < 0) {
+        return -1;
+    }
+    ret = manipulateRestrictApps(appStrUids, chain, restrictAppUids, op);
+    free(chain);
+    return ret;
+}
+
+int BandwidthController::manipulateRestrictApps(const std::vector<std::string>& appStrUids,
+                                                const std::string& chain,
+                                                std::list<int /*appUid*/> &restrictAppUids,
+                                                IptOp op) {
+    for (const auto& appStrUid : appStrUids) {
+        int uid = std::stoi(appStrUid, nullptr, 0);
+        std::list<int /*uid*/>::iterator it;
+        bool isOutputChain = !strncmp(chain.c_str(), "OUTPUT", strlen("OUTPUT"));
+        for (it = restrictAppUids.begin(); it != restrictAppUids.end(); it++) {
+            if (*it == uid)
+                break;
+        }
+        bool found = (it != restrictAppUids.end());
+        if (op == IptOpDelete) {
+            if (!found) {
+                ALOGE("No such appUid %d to remove", uid);
+                return -1;
+            }
+            restrictAppUids.erase(it);
+        } else {
+            if (found && !isOutputChain) {
+                ALOGE("appUid %d exists already", uid);
+                return -1;
+            }
+            restrictAppUids.push_front(uid);
+        }
+    }
+    return manipulateSpecialApps(appStrUids, chain, IptJumpReject, op);
+}
+
 int BandwidthController::manipulateSpecialApps(const std::vector<std::string>& appStrUids,
                                                const std::string& chain, IptJumpOp jumpHandling,
                                                IptOp op) {
diff --git a/server/BandwidthController.h b/server/BandwidthController.h
index 0eef581..75338e6 100644
--- a/server/BandwidthController.h
+++ b/server/BandwidthController.h
@@ -88,6 +88,11 @@ public:
     int addNiceApps(int numUids, char *appUids[]);
     int removeNiceApps(int numUids, char *appUids[]);
 
+    int addRestrictAppsOnData(const char *iface, int numUids, char *appUids[]);
+    int removeRestrictAppsOnData(const char *iface, int numUids, char *appUids[]);
+    int addRestrictAppsOnWlan(const char *iface, int numUids, char *appUids[]);
+    int removeRestrictAppsOnWlan(const char *iface, int numUids, char *appUids[]);
+
     int setGlobalAlert(int64_t bytes);
     int removeGlobalAlert();
     int setGlobalAlertInForwardChain();
@@ -135,6 +140,12 @@ public:
 
     std::string makeDataSaverCommand(IptablesTarget target, bool enable);
 
+    int manipulateRestrictAppsInOut(const char *iface, const std::vector<std::string>& appStrUids,
+                                    std::list<int /*appUid*/> &restrictAppUids, IptOp appOp);
+
+    int manipulateRestrictApps(const std::vector<std::string>& appStrUids, const std::string& chain,
+                               std::list<int /*appUid*/> &restrictAppUids, IptOp appOp);
+
     int manipulateSpecialApps(const std::vector<std::string>& appStrUids, const std::string& chain,
                               IptJumpOp jumpHandling, IptOp appOp);
 
@@ -203,6 +214,9 @@ public:
 
     std::map<std::string, QuotaInfo> mQuotaIfaces;
     std::set<std::string> mSharedQuotaIfaces;
+
+    std::list<int /*appUid*/> restrictAppUidsOnData;
+    std::list<int /*appUid*/> restrictAppUidsOnWlan;
 };
 
 #endif
diff --git a/server/CommandListener.cpp b/server/CommandListener.cpp
index 2795482..cd13f53 100644
--- a/server/CommandListener.cpp
+++ b/server/CommandListener.cpp
@@ -1056,6 +1056,62 @@ int CommandListener::IdletimerControlCmd::runCommand(SocketClient *cli, int argc
         }
         return 0;
     }
+    if (!strcmp(argv[1], "addrestrictappsondata")) {
+        if (argc < 4) {
+            cli->sendMsg(ResponseCode::CommandSyntaxError,
+                       "addrestrictappsondata <interface> <appUid> ...", false);
+            return 0;
+        }
+        if (0 != gCtls->bandwidthCtrl.addRestrictAppsOnData(
+                                        argv[2], argc - 3, argv + 3)) {
+          cli->sendMsg(ResponseCode::OperationFailed, "Firewall command failed", false);
+        } else {
+          cli->sendMsg(ResponseCode::CommandOkay, "Firewall command succeeded", false);
+        }
+        return 0;
+    }
+    if (!strcmp(argv[1], "removerestrictappsondata")) {
+        if (argc < 4) {
+            cli->sendMsg(ResponseCode::CommandSyntaxError,
+                       "removerestrictappsondata <interface> <appUid> ...", false);
+            return 0;
+        }
+        if (0 != gCtls->bandwidthCtrl.removeRestrictAppsOnData(
+                                        argv[2], argc - 3, argv + 3)) {
+          cli->sendMsg(ResponseCode::OperationFailed, "Firewall command failed", false);
+        } else {
+          cli->sendMsg(ResponseCode::CommandOkay, "Firewall command succeeded", false);
+        }
+        return 0;
+    }
+    if (!strcmp(argv[1], "addrestrictappsonwlan")) {
+        if (argc < 4) {
+            cli->sendMsg(ResponseCode::CommandSyntaxError,
+                         "addrestrictappsonwlan <interface> <appUid> ...", false);
+            return 0;
+        }
+        if (0 != gCtls->bandwidthCtrl.addRestrictAppsOnWlan(
+                                        argv[2], argc - 3, argv + 3)) {
+          cli->sendMsg(ResponseCode::OperationFailed, "Firewall command failed", false);
+        } else {
+          cli->sendMsg(ResponseCode::CommandOkay, "Firewall command succeeded", false);
+        }
+        return 0;
+    }
+    if (!strcmp(argv[1], "removerestrictappsonwlan")) {
+        if (argc < 4) {
+            cli->sendMsg(ResponseCode::CommandSyntaxError,
+                         "removerestrictappsonwlan <inteface> <appUid> ...", false);
+            return 0;
+        }
+        if (0 != gCtls->bandwidthCtrl.removeRestrictAppsOnWlan(
+                                        argv[2], argc - 3, argv + 3)) {
+          cli->sendMsg(ResponseCode::OperationFailed, "Firewall command failed", false);
+        } else {
+          cli->sendMsg(ResponseCode::CommandOkay, "Firewall command succeeded", false);
+        }
+        return 0;
+    }
 
     cli->sendMsg(ResponseCode::CommandSyntaxError, "Unknown idletimer cmd", false);
     return 0;
-- 
2.17.0

